File: reconstruct_unified.py
import json
import os
import sys

def reconstruct_unified():
    canon_dir = r"c:\Users\Admin\Desktop\Juegos ENARM\Dr_Swipe_Specs\CORE_CANON"
    index_path = os.path.join(canon_dir, "CORE_INDEX.jsonl")
    data_dir = r"c:\Users\Admin\Desktop\Juegos ENARM\Dr_Swipe_Specs\data"
    output_path = os.path.join(data_dir, "CORE_UNIFIED.jsonl")

    if not os.path.exists(data_dir):
        os.makedirs(data_dir)

    unified_lines = []

    if not os.path.exists(index_path):
        print(f"Error: Index not found at {index_path}")
        return

    with open(index_path, 'r', encoding='utf-8') as f:
        for i, line in enumerate(f):
            if not line.strip(): continue
            try:
                index_entry = json.loads(line)
            except Exception as e:
                print(f"Error parsing index line {i+1}: {e}")
                continue

            domain = index_entry.get('domain', 'unknown')
            topic = index_entry.get('topic', 'unknown')
            rel_path = index_entry.get('path')
            abs_path = os.path.join(canon_dir, rel_path)

            if not os.path.exists(abs_path):
                print(f"Warning: File not found {abs_path}")
                continue

            print(f"Reading {abs_path}...")
            try:
                with open(abs_path, 'r', encoding='utf-8') as jf:
                    content = jf.read().strip()
                    if not content:
                        print(f"Empty file: {abs_path}")
                        continue
                    
                    # Try to parse as JSON first
                    try:
                        data = json.loads(content)
                        if isinstance(data, dict):
                            nodes = data.get('nodes', [])
                        elif isinstance(data, list):
                            nodes = data
                        else:
                            nodes = []
                    except json.JSONDecodeError:
                        # Try to parse as JSONL
                        nodes = []
                        for l in content.splitlines():
                            if l.strip():
                                try:
                                    nodes.append(json.loads(l))
                                except:
                                    pass
                
                for node in nodes:
                    if not isinstance(node, dict):
                        continue
                    
                    # Apply transformation to core_unified_v1
                    unified_node = {
                        "schema": "core_unified_v1",
                        "id": node.get('id'),
                        "domain": domain,
                        "topic": topic,
                        "source_path": rel_path,
                        "source_version": node.get('version'),
                        "tema": node.get('tema'),
                        "subtema": node.get('subtema'),
                        "enunciado": node.get('enunciado') or node.get('prompt') or node.get('pregunta'),
                        "escenario": node.get('escenario'),
                        "tipo_decision": node.get('tipo_decision'),
                        "opciones": [],
                        "respuesta_correcta": node.get('respuesta_correcta'),
                        "justificacion": node.get('justificacion'),
                        "letal_si_falla": node.get('letal_si_falla'),
                        "dificultad": node.get('dificultad'),
                        "tags": node.get('tags'),
                        "fuentes": node.get('fuentes'),
                        "original": node,
                        "extras": {}
                    }

                    # Normalize opciones
                    raw_opciones = node.get('opciones') or node.get('options')
                    if isinstance(raw_opciones, dict):
                        for k, v in raw_opciones.items():
                            resultado = "correcto" if k == node.get('respuesta_correcta') else None
                            unified_node['opciones'].append({
                                "id": k,
                                "texto": v,
                                "resultado": resultado,
                                "impacto": None
                            })
                    elif isinstance(raw_opciones, list):
                        for opt in raw_opciones:
                            unified_node['opciones'].append({
                                "id": opt.get('id'),
                                "texto": opt.get('texto') or opt.get('text'),
                                "resultado": opt.get('resultado') or opt.get('outcome'),
                                "impacto": opt.get('impacto') or opt.get('impact')
                            })

                    unified_lines.append(json.dumps(unified_node, ensure_ascii=False))
            except Exception as e:
                print(f"Error processing {abs_path}: {e}")

    with open(output_path, 'w', encoding='utf-8') as f:
        f.write('\n'.join(unified_lines) + '\n')

    print(f"Successfully created {output_path} with {len(unified_lines)} nodes.")

if __name__ == "__main__":
    reconstruct_unified()

File: reconstruct_unified.js
const fs = require('fs');
const path = require('path');

function reconstructUnified() {
    const canonDir = path.join(__dirname, 'CORE_CANON');
    const indexPath = path.join(canonDir, 'CORE_INDEX.jsonl');
    const dataDir = path.join(__dirname, 'data');
    const outputPath = path.join(dataDir, 'CORE_UNIFIED.jsonl');

    if (!fs.existsSync(dataDir)) {
        fs.mkdirSync(dataDir, { recursive: true });
    }

    if (!fs.existsSync(indexPath)) {
        console.error(`Error: Index not found at ${indexPath}`);
        return;
    }

    const unifiedLines = [];
    const indexContent = fs.readFileSync(indexPath, 'utf-8');
    const indexLines = indexContent.split('\n');

    for (let i = 0; i < indexLines.length; i++) {
        const line = indexLines[i].trim();
        if (!line) continue;

        let indexEntry;
        try {
            indexEntry = JSON.parse(line);
        } catch (e) {
            console.error(`Error parsing index line ${i + 1}:`, e);
            continue;
        }

        const domain = indexEntry.domain || 'unknown';
        const topic = indexEntry.topic || 'unknown';
        const relPath = indexEntry.path;
        if (!relPath) continue;

        const absPath = path.join(canonDir, relPath);

        if (!fs.existsSync(absPath)) {
            console.warn(`Warning: File not found ${absPath}`);
            continue;
        }

        console.log(`Reading ${absPath}...`);
        try {
            const content = fs.readFileSync(absPath, 'utf-8').trim();
            if (!content) {
                console.log(`Empty file: ${absPath}`);
                continue;
            }

            let nodes = [];
            try {
                const data = JSON.parse(content);
                if (data && typeof data === 'object' && !Array.isArray(data)) {
                    nodes = data.nodes || [];
                } else if (Array.isArray(data)) {
                    nodes = data;
                }
            } catch (e) {
                // Try to parse as JSONL
                const fileLines = content.split('\n');
                for (const l of fileLines) {
                    if (l.trim()) {
                        try {
                            nodes.push(JSON.parse(l));
                        } catch (err) {
                            // ignore
                        }
                    }
                }
            }

            for (const node of nodes) {
                if (!node || typeof node !== 'object') continue;

                const unifiedNode = {
                    schema: "core_unified_v1",
                    id: node.id,
                    domain: domain,
                    topic: topic,
                    source_path: relPath,
                    source_version: node.version,
                    tema: node.tema,
                    subtema: node.subtema,
                    enunciado: node.enunciado || node.prompt || node.pregunta,
                    escenario: node.escenario || null,
                    tipo_decision: node.tipo_decision,
                    opciones: [],
                    respuesta_correcta: node.respuesta_correcta || null,
                    justificacion: node.justificacion || null,
                    letal_si_falla: node.letal_si_falla || null,
                    dificultad: node.dificultad || null,
                    tags: node.tags || null,
                    fuentes: node.fuentes || [],
                    original: node,
                    extras: {}
                };

                const rawOpciones = node.opciones || node.options;
                if (rawOpciones) {
                    if (Array.isArray(rawOpciones)) {
                        for (const opt of rawOpciones) {
                            unifiedNode.opciones.push({
                                id: opt.id,
                                texto: opt.texto || opt.text,
                                resultado: opt.resultado || opt.outcome,
                                impacto: opt.impacto || opt.impact
                            });
                        }
                    } else if (typeof rawOpciones === 'object') {
                        for (const [k, v] of Object.entries(rawOpciones)) {
                            const resultado = (k === node.respuesta_correcta) ? "correcto" : null;
                            unifiedNode.opciones.push({
                                id: k,
                                texto: v,
                                resultado: resultado,
                                impacto: null
                            });
                        }
                    }
                }

                unifiedLines.push(JSON.stringify(unifiedNode));
            }
        } catch (e) {
            console.error(`Error processing ${absPath}:`, e);
        }
    }

    fs.writeFileSync(outputPath, unifiedLines.join('\n') + '\n', 'utf-8');
    console.log(`Successfully created ${outputPath} with ${unifiedLines.length} nodes using UTF-8 encoding.`);
}

reconstructUnified();
